File: app/src/main/java/com/yeputra/moviecatalogue/base/BaseActivity.java
package com.yeputra.moviecatalogue.base;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

/**
 * Created by yovi.putra
 *    on 09/Mar/2019 10:56
 * Company SIEMO - PT. Multipolar Technology, Tbk
 */
public abstract class BaseActivity<viewmodel extends IBaseViewModel>
        extends AppCompatActivity
        implements IBaseView {
    protected final String TAG = BaseActivity.class.getSimpleName();

    protected viewmodel viewmodel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        viewmodel = initViewModel();
    }

    @Override
    protected void onDestroy() {
        if(viewmodel != null)
            viewmodel.onDestroy();
        super.onDestroy();
        Log.d(TAG, "Activity destroy");
    }

    protected abstract viewmodel initViewModel();

    @Override
    public Context getContextView() {
        return this;
    }

    @Override
    public void onShowProgressbar() {
        Log.d(TAG, "show progressbar");
    }

    @Override
    public void onHideProgressbar() {
        Log.d(TAG, "hide progressbar");
    }

    @Override
    public void onViewModelFailed(@Nullable String message) {
        Log.d(TAG, "Presenter failed");
    }
}

File: app/src/main/java/com/yeputra/moviecatalogue/base/BaseActivity.kt
package com.yeputra.moviecatalogue.base

import android.content.Context
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity

/**
 * Created by yovi.putra
 * on 09/Mar/2019 10:56
 * Company SIEMO - PT. Multipolar Technology, Tbk
 */
abstract class BaseActivity<viewmodel : IBaseViewModel> : AppCompatActivity(), IBaseView {
    override fun contextView(): Context = this

    private val TAG = BaseActivity::class.java.simpleName

    protected lateinit var viewmodel: viewmodel

    protected abstract fun initViewModel(): viewmodel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewmodel = initViewModel()
    }

    override fun onDestroy() {
        viewmodel?.onDestroy()
        super.onDestroy()
    }

    override fun onPushInformation(message: String?, data: Any?) {}

    override fun onShowProgressbar() {}

    override fun onHideProgressbar() {}

}